File: riveredge-frontend/src/apps/kuaiplm/services/project-proposal.ts
/**
 * 项目建议书 API（R-15 #68）
 */

import { api } from '../../../services/api';

export type SupplierAssessmentLine = {
  material_key: string;
  suppliers_text?: string | null;
};

const BASE = '/apps/kuaiplm/project-proposals';

export type ProjectProposalStatus =
  | 'draft'
  | 'pending'
  | 'approved'
  | 'issued'
  | 'rejected';

export interface ProjectProposal {
  id: number;
  uuid: string;
  proposal_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  title: string;
  summary?: string | null;
  customer_name?: string | null;
  expected_date?: string | null;
  product_lines?: string[];
  proposing_dept?: string | null;
  proposer_name?: string | null;
  proposed_at?: string | null;
  sample_date?: string | null;
  mass_production_date?: string | null;
  sample_quantity?: string | null;
  customer_code?: string | null;
  contact_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  customer_product_model?: string | null;
  customer_material_types?: string[];
  dev_req_types?: string[];
  company_product_model?: string | null;
  cost_change_notes?: string | null;
  supplier_assessment_lines?: SupplierAssessmentLine[];
  procurement_reviewer_name?: string | null;
  supplier_id?: number | null;
  supplier_code?: string | null;
  supplier_name?: string | null;
  supplier_contact?: string | null;
  supplier_remark?: string | null;
  status: ProjectProposalStatus;
  remarks?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  issued_at?: string | null;
  issued_by_name?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface ProjectProposalSalesPayload {
  project_id: number;
  title: string;
  summary?: string | null;
  customer_name?: string | null;
  expected_date?: string | null;
  product_lines?: string[];
  proposing_dept?: string | null;
  proposer_name?: string | null;
  proposed_at?: string | null;
  sample_date?: string | null;
  mass_production_date?: string | null;
  sample_quantity?: string | null;
  customer_code?: string | null;
  contact_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  customer_product_model?: string | null;
  customer_material_types?: string[];
  dev_req_types?: string[];
  company_product_model?: string | null;
  cost_change_notes?: string | null;
  remarks?: string | null;
}

function unwrapList(res: unknown): { items: ProjectProposal[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as ProjectProposal[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const projectProposalApi = {
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    project_id?: number;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => {
    const res = await api.get(`${BASE}/${id}`);
    return res as ProjectProposal;
  },
  create: async (data: ProjectProposalSalesPayload) => {
    const res = await api.post(BASE, data);
    return res as ProjectProposal;
  },
  update: async (id: number, data: Partial<Omit<ProjectProposalSalesPayload, 'project_id'>>) => {
    const res = await api.put(`${BASE}/${id}`, data);
    return res as ProjectProposal;
  },
  fillSupplier: async (
    id: number,
    data: {
      supplier_id?: number | null;
      supplier_code?: string | null;
      supplier_name?: string;
      supplier_contact?: string | null;
      supplier_remark?: string | null;
      supplier_assessment_lines?: SupplierAssessmentLine[];
      procurement_reviewer_name?: string | null;
    },
  ) => {
    const res = await api.put(`${BASE}/${id}/supplier`, data);
    return res as ProjectProposal;
  },
  submit: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/submit`);
    return res as ProjectProposal;
  },
  approve: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/approve`);
    return res as ProjectProposal;
  },
  reject: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/reject`);
    return res as ProjectProposal;
  },
  issue: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/issue`);
    return res as ProjectProposal;
  },
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};

File: riveredge-backend/src/apps/kuaiplm/schemas/project_proposal.py
"""项目建议书 Schema（R-15 #68）"""

from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SupplierAssessmentLine(BaseModel):
    material_key: str = Field(..., min_length=1, max_length=40)
    suppliers_text: Optional[str] = Field(None, max_length=500)


class ProjectProposalCreate(BaseModel):
    project_id: int
    title: str = Field(..., min_length=1, max_length=200)
    proposal_code: Optional[str] = Field(None, max_length=50)
    summary: Optional[str] = None
    customer_name: Optional[str] = Field(None, max_length=200)
    expected_date: Optional[date] = None
    product_lines: List[str] = Field(default_factory=list)
    proposing_dept: Optional[str] = Field(None, max_length=32)
    proposer_name: Optional[str] = Field(None, max_length=100)
    proposed_at: Optional[date] = None
    sample_date: Optional[date] = None
    mass_production_date: Optional[date] = None
    sample_quantity: Optional[str] = Field(None, max_length=80)
    customer_code: Optional[str] = Field(None, max_length=80)
    contact_name: Optional[str] = Field(None, max_length=100)
    contact_phone: Optional[str] = Field(None, max_length=50)
    contact_email: Optional[str] = Field(None, max_length=200)
    customer_product_model: Optional[str] = Field(None, max_length=200)
    customer_material_types: List[str] = Field(default_factory=list)
    dev_req_types: List[str] = Field(default_factory=list)
    company_product_model: Optional[str] = Field(None, max_length=200)
    cost_change_notes: Optional[str] = None
    remarks: Optional[str] = None


class ProjectProposalUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    summary: Optional[str] = None
    customer_name: Optional[str] = Field(None, max_length=200)
    expected_date: Optional[date] = None
    product_lines: Optional[List[str]] = None
    proposing_dept: Optional[str] = Field(None, max_length=32)
    proposer_name: Optional[str] = Field(None, max_length=100)
    proposed_at: Optional[date] = None
    sample_date: Optional[date] = None
    mass_production_date: Optional[date] = None
    sample_quantity: Optional[str] = Field(None, max_length=80)
    customer_code: Optional[str] = Field(None, max_length=80)
    contact_name: Optional[str] = Field(None, max_length=100)
    contact_phone: Optional[str] = Field(None, max_length=50)
    contact_email: Optional[str] = Field(None, max_length=200)
    customer_product_model: Optional[str] = Field(None, max_length=200)
    customer_material_types: Optional[List[str]] = None
    dev_req_types: Optional[List[str]] = None
    company_product_model: Optional[str] = Field(None, max_length=200)
    cost_change_notes: Optional[str] = None
    remarks: Optional[str] = None


class ProjectProposalSupplierFill(BaseModel):
    supplier_id: Optional[int] = None
    supplier_code: Optional[str] = Field(None, max_length=80)
    supplier_name: Optional[str] = Field(None, max_length=200)
    supplier_contact: Optional[str] = Field(None, max_length=200)
    supplier_remark: Optional[str] = None
    supplier_assessment_lines: Optional[List[SupplierAssessmentLine]] = None
    procurement_reviewer_name: Optional[str] = Field(None, max_length=100)


class ProjectProposalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    proposal_code: str
    project_id: int
    project_code: str
    project_name: str
    title: str
    summary: Optional[str] = None
    customer_name: Optional[str] = None
    expected_date: Optional[date] = None
    product_lines: List[str] = Field(default_factory=list)
    proposing_dept: Optional[str] = None
    proposer_name: Optional[str] = None
    proposed_at: Optional[date] = None
    sample_date: Optional[date] = None
    mass_production_date: Optional[date] = None
    sample_quantity: Optional[str] = None
    customer_code: Optional[str] = None
    contact_name: Optional[str] = None
    contact_phone: Optional[str] = None
    contact_email: Optional[str] = None
    customer_product_model: Optional[str] = None
    customer_material_types: List[str] = Field(default_factory=list)
    dev_req_types: List[str] = Field(default_factory=list)
    company_product_model: Optional[str] = None
    cost_change_notes: Optional[str] = None
    supplier_assessment_lines: List[dict] = Field(default_factory=list)
    procurement_reviewer_name: Optional[str] = None
    supplier_id: Optional[int] = None
    supplier_code: Optional[str] = None
    supplier_name: Optional[str] = None
    supplier_contact: Optional[str] = None
    supplier_remark: Optional[str] = None
    status: str
    remarks: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    issued_at: Optional[datetime] = None
    issued_by: Optional[int] = None
    issued_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None


class ProjectProposalListResponse(BaseModel):
    items: List[ProjectProposalResponse]
    total: int
